File: client/internal/invite/invite.go
// Package invite defines the per-user invite blob: a compact, base64url-encoded
// JSON slice of one user's state. It can carry several routers (the user's
// common client_id plus, per router, that router's address, PSK and openable
// services). The admin exports it; the client decodes it and knocks. It carries
// exactly what the runtime needs — no other users' secrets.
package invite

import (
	"encoding/base64"
	"encoding/json"
	"fmt"

	"mikrotik-psk-knock/client/internal/config"
)

// Version is the current blob format version.
const Version = 2

// Blob is one user's runtime configuration across one or more routers.
type Blob struct {
	Version  int      `json:"v"`
	ClientID string   `json:"client_id"` // common identity across routers
	Routers  []Router `json:"routers"`
}

// Router is one router the user can reach: its address, bucket seconds, the PSK
// for this router and the services openable on it.
type Router struct {
	Router        string    `json:"router"` // address
	BucketSeconds int64     `json:"bucket_seconds"`
	PSK           string    `json:"psk"`
	Services      []Service `json:"services"`
}

// Service is one openable service the user has access to.
type Service struct {
	Name      string `json:"name"` // service_name, part of the token
	Stage1    int    `json:"stage1"`
	Stage2    int    `json:"stage2"`
	Token     int    `json:"token"`
	CheckPort int    `json:"check_port"` // external port for post-knock TCP check
}

// Encode marshals the blob to a base64url string.
func Encode(b Blob) (string, error) {
	data, err := json.Marshal(b)
	if err != nil {
		return "", err
	}
	return base64.RawURLEncoding.EncodeToString(data), nil
}

// Decode parses a base64url blob string.
func Decode(s string) (Blob, error) {
	data, err := base64.RawURLEncoding.DecodeString(s)
	if err != nil {
		return Blob{}, fmt.Errorf("invite: base64 decode: %w", err)
	}
	var b Blob
	if err := json.Unmarshal(data, &b); err != nil {
		return Blob{}, fmt.Errorf("invite: json decode: %w", err)
	}
	if b.Version != Version {
		return Blob{}, fmt.Errorf("invite: unsupported version %d (want %d)", b.Version, Version)
	}
	if b.ClientID == "" || len(b.Routers) == 0 {
		return Blob{}, fmt.Errorf("invite: incomplete blob")
	}
	for _, r := range b.Routers {
		if r.Router == "" || r.PSK == "" || len(r.Services) == 0 {
			return Blob{}, fmt.Errorf("invite: incomplete router entry")
		}
	}
	return b, nil
}

// ToConfig converts the blob into a minimal config.Config the runtime can
// resolve against: each blob router becomes a config router (keyed by its
// address) and the user's access to it, with just the ports the client needs.
// The returned config is a runtime-only projection, not meant to pass
// config.Validate.
func (b Blob) ToConfig() config.Config {
	routers := map[string]config.Router{}
	access := map[string]config.UserAccess{}
	for _, rb := range b.Routers {
		services := map[string]config.Service{}
		names := make([]string, 0, len(rb.Services))
		for _, s := range rb.Services {
			services[s.Name] = config.Service{
				ServiceName: s.Name,
				Stage1Port:  s.Stage1,
				Stage2Port:  s.Stage2,
				TokenPort:   s.Token,
				Target:      config.Target{Port: s.CheckPort},
			}
			names = append(names, s.Name)
		}
		routers[rb.Router] = config.Router{
			Address:  rb.Router,
			Defaults: config.Defaults{BucketSeconds: rb.BucketSeconds},
			Services: services,
		}
		access[rb.Router] = config.UserAccess{Services: names, PSK: rb.PSK}
	}
	return config.Config{
		Routers: routers,
		Users:   map[string]config.User{b.ClientID: {ClientID: b.ClientID, Access: access}},
	}
}

// RouterNames returns the blob's router addresses (keys) in slice order.
func (b Blob) RouterNames() []string {
	names := make([]string, 0, len(b.Routers))
	for _, r := range b.Routers {
		names = append(names, r.Router)
	}
	return names
}

File: client-macos/Sources/MkpkKit/Invite.swift
import Foundation

/// The per-user invite blob — a base64url-encoded JSON matching the Go
/// `internal/invite` v2 format. It carries only what a client needs to knock:
/// the public router address, the PSK, and each service's ports.
public struct Invite: Codable, Sendable, Equatable {
    public let version: Int
    public let clientID: String
    public let routers: [RouterInvite]

    enum CodingKeys: String, CodingKey {
        case version = "v"
        case clientID = "client_id"
        case routers
    }
}

public struct RouterInvite: Codable, Sendable, Equatable {
    /// Public address end users knock (the only address in the blob).
    public let router: String
    public let bucketSeconds: Int64
    public let psk: String
    public let services: [ServiceInvite]

    enum CodingKeys: String, CodingKey {
        case router
        case bucketSeconds = "bucket_seconds"
        case psk
        case services
    }
}

public struct ServiceInvite: Codable, Sendable, Equatable {
    /// service_name — part of the token formula.
    public let name: String
    public let stage1: Int
    public let stage2: Int
    public let token: Int
    /// External port for the post-knock TCP check (0 → no auto-check possible).
    public let checkPort: Int
    /// Effective TTL of the allowed entry after a knock (e.g. "3m"), so the UI can
    /// show how long the port stays open. Informational; the router is authority.
    /// Absent on invites from older exporters.
    public let allowedTimeout: String?

    enum CodingKeys: String, CodingKey {
        case name, stage1, stage2, token
        case checkPort = "check_port"
        case allowedTimeout = "allowed_timeout"
    }
}

public enum InviteError: Error, Equatable {
    case notBase64URL
    case decode(String)
}

extension Invite {
    /// Decode an invite from its base64url blob string (as produced by
    /// `mkpk-provision export`). Whitespace/newlines are trimmed.
    public static func decode(blob: String) throws -> Invite {
        let trimmed = blob.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let data = base64URLDecode(trimmed) else {
            throw InviteError.notBase64URL
        }
        do {
            return try JSONDecoder().decode(Invite.self, from: data)
        } catch {
            throw InviteError.decode(String(describing: error))
        }
    }

    /// Read an invite from a file (a `.mkpk` produced by export).
    public static func load(contentsOf url: URL) throws -> Invite {
        let text = try String(contentsOf: url, encoding: .utf8)
        return try decode(blob: text)
    }
}

/// Decode base64url (RFC 4648 §5, no padding) into Data, tolerating optional
/// padding. Returns nil on invalid input.
func base64URLDecode(_ s: String) -> Data? {
    var str = s.replacingOccurrences(of: "-", with: "+")
        .replacingOccurrences(of: "_", with: "/")
    let remainder = str.count % 4
    if remainder > 0 {
        str += String(repeating: "=", count: 4 - remainder)
    }
    return Data(base64Encoded: str)
}
